app/src/heartbeat/heartbeat_storage_desktop.cc

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "app/src/heartbeat/heartbeat_storage_desktop.h"
+
+#include <fstream>
+#include <regex>
+#include <vector>
+
+#include "app/logged_heartbeats_generated.h"
+#include "app/src/filesystem.h"
+#include "app/src/logger.h"
+
+namespace firebase {
+namespace heartbeat {
+
+using LoggedHeartbeatsFlatbuffer =
+    com::google::firebase::cpp::heartbeat::LoggedHeartbeats;
+using com::google::firebase::cpp::heartbeat::CreateLoggedHeartbeats;
+using com::google::firebase::cpp::heartbeat::CreateUserAgentAndDates;
+using com::google::firebase::cpp::heartbeat::GetLoggedHeartbeats;
+using com::google::firebase::cpp::heartbeat::UserAgentAndDates;
+using com::google::firebase::cpp::heartbeat::VerifyLoggedHeartbeatsBuffer;
+
+// Using an anonymous namespace for helper to construct filename
+namespace {
+
+const char kHeartbeatDir[] = "firebase-heartbeat";
+const char kHeartbeatFilenamePrefix[] = "heartbeats-";
+
+std::string CreateFilename(const std::string& app_id, Logger* logger) {
+  std::string error;
+  std::string app_dir =
+      AppDataDir(kHeartbeatDir, /*should_create=*/true, &error);
+  if (!error.empty()) {
+    logger->LogError(error.c_str());
+    return "";
+  }
+  if (app_dir.empty()) {
+    return "";
+  }
+
+  // Remove any symbols from app_id that might not be allowed in filenames.
+  auto app_id_without_symbols =
+      std::regex_replace(app_id, std::regex("[/\\\\?%*:|\"<>.,;=]"), "");
+  // Note: fstream will convert / to \ if needed on windows.
+  return app_dir + "/" + kHeartbeatFilenamePrefix + app_id_without_symbols;
+}
+
+}  // namespace
+HeartbeatStorageDesktop::HeartbeatStorageDesktop(const std::string& app_id,
+                                                 Logger* logger)
```

Comment:
If `logger` is optional, you should do null-check whenever using it.

An easy solution is to make it required by using `const Logger& logger` and keep a reference to it.
Or you can implement functions like 
```
void HeartbeatStorageDesktop::LogWarning(const char* format, ...) {
if (logger_) {
  va_list list;
  va_start(list, format);
  logger_->LogWarning(format, list)
  va_end(list);
}
}
```
See  this https://github.com/firebase/firebase-cpp-sdk/blob/main/app/src/logger.cc#L23.

But I think this could be a bit too much trouble.